url7f729337e3_timmolter_Sundial_tgz-pJ8-org_knowm_sundial_SampleRunJ8/src/org/quartz/QuartzScheduler.java

```diff
@@ -0,0 +1,1227 @@
+/**
+ * Copyright 2015 Knowm Inc. (http://knowm.org) and contributors.
+ * Copyright 2013-2015 Xeiam LLC (http://xeiam.com) and contributors.
+ * Copyright 2001-2011 Terracotta Inc. (http://terracotta.org).
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.quartz;
+
+import static org.quartz.builders.SimpleTriggerBuilder.simpleTriggerBuilder;
+
+import java.util.ArrayList;
+import java.util.Date;
+import java.util.HashMap;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.Random;
+import java.util.Set;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import org.quartz.classloading.CascadingClassLoadHelper;
+import org.quartz.core.Calendar;
+import org.quartz.core.JobExecutionContext;
+import org.quartz.core.QuartzSchedulerResources;
+import org.quartz.core.QuartzSchedulerThread;
+import org.quartz.core.Scheduler;
+import org.quartz.core.SchedulerSignaler;
+import org.quartz.core.SchedulerSignalerImpl;
+import org.quartz.exceptions.JobExecutionException;
+import org.quartz.exceptions.JobPersistenceException;
+import org.quartz.exceptions.ObjectAlreadyExistsException;
+import org.quartz.exceptions.SchedulerException;
+import org.quartz.jobs.InterruptableJob;
+import org.quartz.jobs.Job;
+import org.quartz.jobs.JobDataMap;
+import org.quartz.jobs.JobDetail;
+import org.quartz.jobs.JobFactory;
+import org.quartz.jobs.SimpleJobFactory;
+import org.quartz.listeners.JobListener;
+import org.quartz.listeners.ListenerManager;
+import org.quartz.listeners.ListenerManagerImpl;
+import org.quartz.listeners.SchedulerListener;
+import org.quartz.listeners.SchedulerListenerSupport;
+import org.quartz.listeners.TriggerListener;
+import org.quartz.plugins.SchedulerPlugin;
+import org.quartz.triggers.OperableTrigger;
+import org.quartz.triggers.Trigger;
+import org.quartz.triggers.Trigger.CompletedExecutionInstruction;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * <p>
+ * This is the heart of Quartz, an indirect implementation of the <code>{@link org.quartz.core.Scheduler}</code> interface, containing methods to
+ * schedule <code>{@link org.quartz.jobs.Job}</code>s, register <code>{@link org.quartz.listeners.JobListener}</code> instances, etc.
+ * </p>
+ *
+ * @author James House
+ */
+public class QuartzScheduler implements Scheduler {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

This class references 47 other classes. By comparison, 98% of the classes in the CodeGuru reference dataset reference fewer. This indicates that this class is highly coupled with other classes. A class that is highly coupled with other classes is difficult to understand and its behavior might change unexpectedly when one of its referenced classes is updated. High coupling could also increase the integration test complexity, maintenance cost and technical debt. We recommend that you simplify this class or break it into multiple classes.